File: backend/app/services/attendance_service.py
from datetime import date, datetime

from app.utils.tenant import get_db
from app.models.attendance import Attendance
from app.models.student_section import StudentSection


def _parse_date(date_str: str) -> date:
    return datetime.strptime(date_str, '%Y-%m-%d').date()


class AttendanceService:

    @staticmethod
    def mark_attendance(section_id: int, date_str: str, records: list, marked_by_user_id: int):
        """
        Bulk-mark attendance for all students in a section for one date.

        records: [{"student_id": int, "status": str}, ...]
        Returns (result_dict, None) or (None, error_dict).
        409 if any row already exists for this section + date.
        """
        target_date = _parse_date(date_str)

        existing = (
            get_db()
            .query(Attendance)
            .filter_by(section_id=section_id, date=target_date)
            .first()
        )
        if existing:
            return None, {
                'message': f'Attendance already marked for section {section_id} on {date_str}',
                'status': 409,
            }

        created = []
        for rec in records:
            row = Attendance(
                student_id=rec['student_id'],
                section_id=section_id,
                date=target_date,
                status=rec['status'],
                marked_by=marked_by_user_id,
            )
            get_db().add(row)
            created.append(row)

        get_db().commit()

        # Fire absence notifications after commit
        from app.services.notification_service import NotificationService
        for row in created:
            if row.status == 'absent':
                NotificationService.notify_absence(row.student_id, target_date)

        return {
            'section_id': section_id,
            'date': date_str,
            'records_saved': len(created),
        }, None

    @staticmethod
    def get_for_student(student_id: int, month: int, year: int):
        """Return all attendance rows for a student in the given month/year."""
        import calendar
        _, last_day = calendar.monthrange(year, month)
        from_date = date(year, month, 1)
        to_date = date(year, month, last_day)

        rows = (
            get_db()
            .query(Attendance)
            .filter(
                Attendance.student_id == student_id,
                Attendance.date >= from_date,
                Attendance.date <= to_date,
            )
            .order_by(Attendance.date)
            .all()
        )
        return [r.to_dict() for r in rows]

    @staticmethod
    def get_report(section_id: int, from_date_str: str, to_date_str: str):
        """
        Aggregate attendance for a section over a date range.
        Returns per-student summary + per-day counts.
        """
        from_date = _parse_date(from_date_str)
        to_date = _parse_date(to_date_str)

        rows = (
            get_db()
            .query(Attendance)
            .filter(
                Attendance.section_id == section_id,
                Attendance.date >= from_date,
                Attendance.date <= to_date,
            )
            .order_by(Attendance.date, Attendance.student_id)
            .all()
        )

        student_summary: dict = {}
        for row in rows:
            sid = row.student_id
            if sid not in student_summary:
                student_summary[sid] = {'student_id': sid, 'present': 0, 'absent': 0,
                                        'late': 0, 'leave': 0, 'holiday': 0}
            student_summary[sid][row.status] = student_summary[sid].get(row.status, 0) + 1

        return {
            'section_id': section_id,
            'from_date': from_date_str,
            'to_date': to_date_str,
            'total_records': len(rows),
            'student_summaries': list(student_summary.values()),
        }

    @staticmethod
    def get_today_summary():
        """Count of each status value across all sections for today."""
        today = date.today()
        rows = (
            get_db()
            .query(Attendance)
            .filter(Attendance.date == today)
            .all()
        )
        summary = {'present': 0, 'absent': 0, 'late': 0, 'leave': 0, 'holiday': 0}
        for row in rows:
            summary[row.status] = summary.get(row.status, 0) + 1
        summary['total'] = len(rows)
        summary['date'] = str(today)
        return summary

File: frontend/src/app/core/services/attendance.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { ApiResponse } from '../models/user.model';

// ── Shared interfaces ──────────────────────────────────────────────────────

export type AttendanceStatus = 'present' | 'absent' | 'late' | 'leave' | 'holiday';

export interface AttendanceRecord {
  id: number;
  student_id: number;
  section_id: number;
  date: string;
  status: AttendanceStatus;
  remarks?: string | null;
  marked_by?: number | null;
  created_at?: string;
  updated_at?: string;
}

export interface AttendanceMarkEntry {
  student_id: number;
  status: AttendanceStatus;
}

export interface AttendanceMarkPayload {
  section_id: number;
  date: string;
  records: AttendanceMarkEntry[];
}

export interface AttendanceMarkResponse {
  section_id: number;
  date: string;
  records_created: number;
}

export interface StudentAttendanceData {
  attendance: AttendanceRecord[];
}

export interface StudentSummary {
  student_id: number;
  student_name?: string;
  admission_no?: string;
  present: number;
  absent: number;
  late: number;
  leave: number;
  holiday: number;
  total: number;
}

export interface AttendanceReportData {
  section_id: number;
  from_date: string;
  to_date: string;
  total_records: number;
  student_summaries: StudentSummary[];
}

export interface TodaySummaryData {
  present: number;
  absent: number;
  late: number;
  leave: number;
  holiday: number;
  total: number;
  date: string;
}

// ── Service ────────────────────────────────────────────────────────────────

@Injectable({ providedIn: 'root' })
export class AttendanceService {
  private readonly apiUrl = '/api/v1/attendance';

  constructor(private http: HttpClient) {}

  markAttendance(payload: AttendanceMarkPayload): Observable<ApiResponse<AttendanceMarkResponse>> {
    return this.http.post<ApiResponse<AttendanceMarkResponse>>(`${this.apiUrl}/mark`, payload);
  }

  getAttendance(studentId: number, month: number, year: number): Observable<ApiResponse<StudentAttendanceData>> {
    const params = new HttpParams()
      .set('student_id', studentId.toString())
      .set('month', month.toString())
      .set('year', year.toString());
    return this.http.get<ApiResponse<StudentAttendanceData>>(this.apiUrl, { params });
  }

  getReport(sectionId: number, fromDate: string, toDate: string): Observable<ApiResponse<AttendanceReportData>> {
    const params = new HttpParams()
      .set('section_id', sectionId.toString())
      .set('from_date', fromDate)
      .set('to_date', toDate);
    return this.http.get<ApiResponse<AttendanceReportData>>(`${this.apiUrl}/report`, { params });
  }

  getTodaySummary(): Observable<ApiResponse<TodaySummaryData>> {
    return this.http.get<ApiResponse<TodaySummaryData>>(`${this.apiUrl}/today-summary`);
  }
}
